### ISSUE TITLE
Player visual angle issues

### ISSUE THREAD

**Issue by EnceladusX2** [Techokami/SonicWorldsNext#195]
When moving along any sloped surface that would change the player's visual angle, their sprite rotation tends to reset to 0 at what feels like random intervals. It's definitely not intended behavior.

https://github.com/user-attachments/assets/e71cfc7d-3158-4e68-a4de-4f3720d37ee3

**Comment by Caverns4** [comment]
I'll try to look into this this weekend.

**Comment by EnceladusX2** [comment]
Did some testing and seem to have boiled this down to the angle collected from collision normals being slightly inconsistent. Logically that makes sense, but with the way this framework handles sprite rotation that's not good. It might be worth just reworking sprite rotation instead of rewriting how the ground angle is collected.

**Comment by Caverns4** [comment]
So perhaps using a lerp value might be better?

**Comment by DimensionWarped** [comment]
If you watch this while paying close attention to the individual tiles and where they connect specifically with the player's collision box, I think you'll get a feel for what's happening here. The player's angle is being calculated by the normal of the contacted tile. For some of these tiles, the angle is more or less severe than it appears it should be relative to the other tiles. They hover right around the transition point between flat mode and 45 degree mode, so those minor angle differences result in a full shift.

I have some ideas on how to smooth this out, but it also kind of exposed a big problem with certain transitions from upward angles to downward angles which you can see towards the end of the video.

https://github.com/user-attachments/assets/932e8ec1-5548-428f-aa9d-7da2e2323f46

**Comment by DimensionWarped** [comment]
I've added the ability to use 'smooth rotation' as an option in the main branch (which now uses 4.3, so don't switch to it unless you are ready to move to Godot 4.3). This doesn't address the issue, so we are leaving it open... I have plans to include an optional 'visual angle' property to tiles that only controls which angle the sprite rotates at when interacting with the tile to deal with this.

**Comment by NeonSRB2** [comment]
Well that was a relief... but i don't think thats a good solution. This is
very well an issue with the tiles themselves.

**Comment by DimensionWarped** [comment]
I agree, what I did isn't a fix at all. It's just a useful diagnostic tool for you to figure out why it's happening and optionally to just use in favor of cleaning up your tiles to not run into that problem.

The options to actually fix it are either what I said above (explicit angle settable per tile that controls the visual angle) or to do something completely different.

It might be that the most proper fix is to make it to where it only visually snaps from one angle to another when it's at the 60% point. IE

```
from -> to      requires
0       -> 45     27 (instead of >22.5)
45     -> 0       18 (instead of < 22.5)
45     -> 90     72 (instead of > 112.5)
90     -> 45     63 (instead of < 112.5)
``` 

in that way, you become much less likely to snap back and forth because of small angle changes between tiles... though it does change the steepness you have to be at to get *any* change. I might experiment with smaller differences too like 55%

**Comment by NeonSRB2** [comment]
I was *trying* to say that I don't think forcing a certain visual angle or
changing when that angle is reached is a good idea because the original
games never did it, unless it only needs to be extended before where the
original byte angle format would be able to store the angle.
Unless they did, then I'm just dumb

**Comment by DimensionWarped** [comment]
I agree with that point, I don't think they did either since they wouldn't need to... they just manually specified the angle on every tile. That's what my other point was about... we *can* bake an override angle into the tileset. I just think that's also probably going to be a huge pain and worse, hard to explain. Not that the current state of things is easy to explain either... I guess maybe the best starting point would just be to clean up the existing tiles, but I really don't enjoy that kind of work.

**Comment by Candescence** [comment]
Just noting that Neon's comments are _incomprehensible_ in the email notifications lmao

But I do legit wonder if Godot 4.5's new tilemap collision merging (aka merging the tile collisions into larger colliders) affects this issue any? I know there's still a transition to 4.3 (though I've been using Worlds Next in 4.4 with no real issues) and it might be something worth investigating despite 4.5 still being in early dev state.

**Comment by DimensionWarped** [comment]
I would say that it *shouldn't*, though I don't know with certainty. Honestly I've got all kinds of questions about how that system works when the tiles themselves can include metadata... if you make a big collision shape does it then check another collision for the smaller tiles if it finds out that you are colliding with a combined tile? Either way though, the whole point of the feature is that the shape of the tile map isn't supposed to change, so the normal angle of the piece of the tilemap you are colliding with should be the same.

**Comment by DimensionWarped** [comment]
The current branch I'm working on fixes this spot at the tileset collision level, so no code changes for that one... though I do need to warn everyone that my current branch has a lot of changes and is probably at least a week out before I can merge it. Oh and it also adds a new debug tool that allows you to toggle the time scale to be slow (1/8th normal specifically) and back which I personally found useful for analyzing these problems.

Quick word of warning though that fixing this one hill means this problem is almost certainly expressed elsewhere. I'm kind of inclined to add a new debug tool that lets you click on a tile to get a quick rundown of possible normal angles for it for the four modes... but that honestly sounds a little tricky to me.

**Comment by Candescence** [comment]
> I would say that it _shouldn't_, though I don't know with certainty. Honestly I've got all kinds of questions about how that system works when the tiles themselves can include metadata... if you make a big collision shape does it then check another collision for the smaller tiles if it finds out that you are colliding with a combined tile? Either way though, the whole point of the feature is that the shape of the tile map isn't supposed to change, so the normal angle of the piece of the tilemap you are colliding with should be the same.

IMO a coordinate-based system based on the collision point makes the most sense as far as I'm concerned.

**Comment by DimensionWarped** [comment]
I'm not sure I follow what that means. Draw a picture maybe?

**Comment by Candescence** [comment]
I mean basically figure out which tile is being hit based on which 'grid square' the collision position is in. The system would just need to know which tile layer the collision happened on and at that point determining which tile was collided with would be straightforward. The collision point should theoretically always be within the tile's grid space, even for tiles that have collisions that take up the entire space, unless that's not how that actually works in Godot.

**Comment by DimensionWarped** [comment]
We already have the ability to know what tile a ray is hitting. The issue isn't know what tile we are hitting, it's that collision has to be hand painted onto each tile in such a way as to avoid normal angles accidentally getting fudged going up or down certain arrangements. One of the approaches I was describing before would be to embed some data in the tiles to describe what angle you should visually present the player as depending on which Ceiling/Wall/Floor mode they are in and then use that to override the angle you would calculate based on the normal if that data is present. But I don't really think we *want* to do that because it complicates the process of making tilesets.

For now I'm just going to focus on cleaning up instances of this from the example tileset where we find them and also provide as many tools as I can for helping to make finding the iffy spots easier.

**Comment by Caverns4** [comment]
One thing we could do is just add a variable that's called an "apparent_angle", and have that move_toward(apparent_angle,angle,delta) at a fixed rate like delta*60, and have the sprite rotation update based on that. Doing that in Sonic 2 Retold produced way better looking results in a couple areas where Sonic used to "twitch" as well.

**Comment by DimensionWarped** [comment]
I'm going to go ahead and close this for now since for the most part I think we've cleared up the individual tiles that were causing this and it's not *really* an engine problem as much as it is a level design problem. I talked about it with Sharb a while back and we both agree we don't really want to deviate from the way the main engine handles terrain angle just because there is some potential for messy tiles incurred by level design. If I feel like it later, I might add the ability for tiles to set their own overrides, but no promises.
